### DISCUSSION TITLE
Why nanobind can build this binding while pybind11 raises "use of undefined type"?

### DISCUSSION THREAD

**Discussion by davidlatwe** [wjakob/nanobind#499] [Q&A]
## Problem

I found this when making a Python binding for [ImGui - docking branch](https://github.com/ocornut/imgui/tree/docking).

The problem starts from here, ImGui's `std::vector<>`-like class, `sizeof()` is used:
https://github.com/ocornut/imgui/blob/085781f5ca5372d5fc804d7e44b5bf27a8994af7/imgui.h#L2003
```c++
// imgui.h
template<typename T>
struct ImVector {
    inline ImVector<T>& operator=(const ImVector<T>& src)   { clear(); resize(src.Size); if (src.Data) memcpy(Data, src.Data, (size_t)Size * sizeof(T)); return *this; }
}
```

In its internal header, `ImVector` is used for storing `ImGuiDockNodeSettings`.
https://github.com/ocornut/imgui/blob/085781f5ca5372d5fc804d7e44b5bf27a8994af7/imgui_internal.h#L1881
```c++
// imgui_internal.h
struct ImGuiDockContext {
    ImVector<ImGuiDockNodeSettings> NodesSettings;
};
```

But the definition of `ImGuiDockNodeSettings` is in `imgui.cpp` file:
https://github.com/ocornut/imgui/blob/085781f5ca5372d5fc804d7e44b5bf27a8994af7/imgui.cpp#L15630

So when I build the binding with pybind11, Windows is complaining:
```
error C2027: use of undefined type 'ImGuiDockNodeSettings'
```
(strange that MSVC did not mention anything about `sizeof` and I thought it was becasue my cmake setup was incorrect.)

And on Linux I saw:
```
error: invalid application of ‘sizeof’ to incomplete type ‘ImGuiDockNodeSettings’
```

## Question

However, if I build the binding with nanobind, build was success and the binding works perfectly.

What magic did nanobind use? 

I made a minimum reproducible repo for showcasing the issue: https://github.com/davidlatwe/py-binding-test

For my own ImGui binding work, I may tweak ImGui source code for pybind11, or just continue using nanobind. But I am really curious about "why this works".

P.S. The title of this thread feels not clear enough, welcome to change it.

Thanks.

**Comment by wjakob** [comment]
Sorry, I don't see how any of this is related to nanobind based on your description.

**Reply by davidlatwe** [reply]
Thank you for your reply. I'm investigating why pybind11 and nanobind yield different results when compiling the same binding code. While I understand that pybind11 is header-only and nanobind uses a support library (e.g., nanobind-static.lib), there seems to be more to it.

In my example, this is the binding code that compiles differently with nanobind and pybind11:
```c++
#ifdef USE_NANOBIND
#   include <nanobind/nanobind.h>
#   define ADD_MODULE NB_MODULE
#else
#   include <pybind11/pybind11.h>
#   define ADD_MODULE PYBIND11_MODULE
#endif

ADD_MODULE(example, m /*'m' for module*/) {
    m.def("GetContext", GetContext);
}
```

The function `GetContext` is returns a pointer to a `Context` struct:
```c++
struct Context {
    Vector<Node> Nodes;
};

Context* GetContext();
```

In this structure, the `Context` member `Nodes` is a template container `Vector` that holds Node objects.

The issue arises with the definition of `Node`, which is located in a `.cpp` file. When compiling with pybind11, the compiler raises an "undefined type" error, which does not occur with nanobind.

Could this discrepancy be due to differences in how nanobind and pybind11 handle template visibility or type linking? I'd appreciate any insights or suggestions on this matter.
